### ISSUE TITLE
Provide separate crate for types only.

### ISSUE THREAD

**Issue by SirWindfield** [Keats/tera#554]
During the development of https://github.com/SirWindfield/tera-text-filters I noticed that depending on `tera` takes up some compilation time that could be avoided when creating "filter/test/function crates". It might be a nice goal to split off Tera's types and utility functions into a separate crate that can be used by third parties to implement filters.

API that might be relevant to split off:
- `Result`
- `Value` and `to_value`
- `try_get_value`

Additionally, it might be a good idea to add a trait called `FilterRegistry` or something similar. That way, functions like https://github.com/SirWindfield/tera-text-filters/blob/f45d22969af7a36b8cb8595cfdeed59bc80202da/src/lib.rs#L8-L16 could be implemented without having to depend on `tera` itself.
The `Tera` struct would then implement the trait that could look like this:
```rust
pub trait FilterRegistry {
    fn register_filter<F: Filter + 'static>(&mut self, name: &str, filter: F);
}
```

An extension of this would be to either combine it together with functions, offering one single trait `Registry` or split this into separate traits and crates:
- `tera-types` that exports the above mentioned API (Result, Value, ...)
- `tera-filter-api` that contains the `FilterRegistry` and the needed `Filter` trait.
- `tera-function-api` that contains the `FunctionRegistry` and the needed `Function` trait.
- `tera-tester-api` that contains the `TesterRegistry` and the needed `Tester` trait.

What do you think? Is that something that you would agree on? This would reduce compilation time quite substantially, people would probably only have to compile serde-json and its dependencies alongside the traits that are put into those api crates. A side-effect would also be that compilation time of `tera` itself would be lowered as the above three crates can be compiled in parallel now.

Would you be open for a PR that does this split? If yes, I'd probably do this using a cargo workspace.

**Comment by Keats** [comment]
I really don't want to split the crate. I expect 99% of the users of Tera are not going to pull additional filters/etc from crates.io but rather implement what they need directly in code. If they do end up pulling a crate like yours, it will only be compiled once anyway provided there are no changes in the major versions.
